compat/src/suspense-list.js

```diff
@@ -0,0 +1,72 @@
+import { Component, options } from 'preact';
+import { Suspense } from './suspense';
+
+// Hook for Suspense boundaries to ask for any extra work before rendering suspended children.
+options.__onSuspensionComplete = (vnode, cb) => {
+	if (vnode._parent._component && vnode._parent._component.__modifySuspense) {
+		vnode._parent._component.__modifySuspense(vnode, cb);
+	} else {
+		cb();
+	}
+};
+
+// having custom inheritance instead of a class here saves a lot of bytes
+export function SuspenseList(props) {
+	// Todo(prateekbh): use a better name before sending the PR.
+	this._thrillers = [];
+}
+
+// Things we do here to save some bytes but are not proper JS inheritance:
+// - call `new Component()` as the prototype
+// - do not set `Suspense.prototype.constructor` to `Suspense`
+SuspenseList.prototype = new Component();
+
+SuspenseList.prototype.__modifySuspense = function(vnode, cb) {
+	switch (this.props.revealOrder) {
+		case 'forwards':
+		case 'backwards':
+			/**
+			 * Forwards and backwards work the same way.
+			 * The direction is controlled in render method while creating `_thrillers` itself.
+			 */
+			if (this._thrillers && this._thrillers[0].vnode === vnode) {
+				cb();
+				this._thrillers.shift();
+				/**
+				 * Find and execute all callbacks in order from 2nd position.
+				 * Breaks as soon as a non resolved(cb===null) suspense found.
+				 */
+				this._thrillers.find(thrill => {
```

Comment:
Unfortunately, [`find` isn't in our browser matrix](https://kangax.github.io/compat-table/es5/) 😢 since we don't compile to ES6. However, you should be able to emulate this `find` call with `.some` and use the return value to break out of the loop.

For the others you can use `.some` and modify the first matching object in your callback when you find the match and then return `true` to break.

Reply:
I chatted with @prateekbh about a parallel approach. If the `_thrillers` array's size is `n` then the worst-case amount of work is something like O(n^2), and some work could be avoided by using a singly linked list and a `Map`. Here's an incomplete sketch, sorry about that 🙂

```js
export function SuspenseList(props) {
  this._head = null;
  this._vnodes = null;
}

SuspenseList.prototype.render = function(props, state) {
  // Points to the first node of a linked list storing the callbacks,
  // where the first node points to the next callback that should be called.
  this._head = null;
  // Maps vnodes to nodes in the linked list.
  this._vnodes = new Map();

  // Build the linked list by iterating through the children in a reverse order
  // (so that _head points to the first node in the end).
  const children = props.children.filter(
    child => child.type.name === Suspense.name
  );
  if (props.revealOrder === "forwards") {
    children.reverse();
  }
  children.some(vnode => {
    this._vnodes.set(vnode, (this._head = { cb: null, next: this._head }));
  });
  return props.children;
};

SuspenseList.prototype.__modifySuspense = function(vnode, cb) {
  const node = this._vnodes.get(vnode);
  node._vnodes.delete(vnode);
  node.cb = cb;

  const order = this.props.revealOrder;
  if (order === "forwards" || order === "backwards") {
    if (node === this._head) {
      consume(this);
    }
  } else if (order === "together") {
    if (this._vnodes.size === 0) {
      consume(this);
    }
  } else {
    cb();
  }
};

// A helper to walk down and consume the linked list as long as there are resolved callbacks.
const consume = list => {
  for (let node; (node = list._head) && node.cb; list._head = node.next) {
    node.cb();
  }
};
```